### ISSUE TITLE
Indentation rule ignored for unknown nodes

### ISSUE THREAD

**Issue by yhpark** [eslint/eslint#9975]
**Tell us about your environment**

* **ESLint Version:** v4.17
* **Node Version:** latest
* **npm Version:** latest

**What parser (default, Babel-ESLint, etc.) are you using?** babel-eslint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

`parser: "babel-eslint"` with `indent` rule enabled (e.g. space, 2).

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
class A {
 foo = 1;
       @SomeDecorator()
     bar;
}
```

```bash
eslint
```

**What did you expect to happen?**
`indent` rule should emit indentation error for all three cases.

**What actually happened? Please include the actual, raw output from ESLint.**
No error detected. Although babel-eslint parses the code correctly, the node types are unknown to `indent` rule so their indentations are just ignored.

**Comment by not-an-aardvark** [comment]
Hi,

This is working as designed. The `indent` rule can't determine the appropriate indentation for node types that it doesn't know about, so it ignores those nodes to avoid reporting a large number of unwanted errors.

You might be interested in [this gist](https://gist.github.com/not-an-aardvark/59fdfed4df22c12cfd3ab6ebf0b1e760) from when the current behavior was designed, which gives an overview of the possible alternatives and explains some of the advantages of the current approach.

**Comment by yhpark** [comment]
Thanks for the response. I just read the gist and yet I don't understand why the approach 2 might not be feasible. Can't we treat unknown nodes like comments? What kind of unwanted error might occur with this approach?

**Comment by yhpark** [comment]
If the new node cannot be multiline, or if the new node is composed of the nodes that are already known, wouldn't it be possible to provide indentation rule easily by just adding them into `KNOWN_NODES`? For example when `eslint-babel` parses ClassProperty and Decorators, these nodes are further parsed into the known node types such as identifiers and punctuators.

I just created a custom indent rule by monkey patching the existing indent rule to avoid ignoring the new node types `ClassProperty` and `Decorator`: https://github.com/yhpark/eslint-plugin-babel/commit/136e1e65f957e61c31f6a8c131970a1e0fd8cc96, and it seems to work well 🤔.
If this makes sense, it would be great if there's an option to provide additional 'known' nodes to apply default indent behavior for them (by not ignoring them).

**Comment by not-an-aardvark** [comment]
Please don't monkeypatch rules. Copy-paste them if you need to, but monkeypatching just means that your code will break if we refactor the implementation.

In general, we don't add special logic for experimental node types (such as `ClassProperty` or `Decorator`) to core rules, because we only support stage 4 features. However, the rule is designed so that rules from plugins would be able to check the indentation of custom nodes without needing to deal with the rest of the semantics of the `indent` rule.

**Comment by platinumazure** [comment]
Would it be possible to add an option for unknown node types which basically defines tokens that should trigger indent or deindent on the same or next line? For example, even if we don't know about a node type, we should be able to let a user specify that curly braces should be used to increase/decrease indent, so they get similar behavior as an if statement.

**Comment by not-an-aardvark** [comment]
It would be possible to do that, but I'm not convinced it would be worth the complexity. I think it would be better for people to use custom rules to enforce this.

**Comment by not-an-aardvark** [comment]
Closing because this is working as intended.
